libvast/vast/system/local_segment_store.hpp

```diff
@@ -0,0 +1,82 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2021 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#include "vast/fwd.hpp"
+
+#include "vast/chunk.hpp"
+#include "vast/segment_builder.hpp"
+#include "vast/system/actors.hpp"
+
+#include <caf/typed_event_based_actor.hpp>
+
+namespace vast::system {
+
+/// The STORE BUILDER actor interface.
+using local_store_actor
+  = typed_actor_fwd<caf::reacts_to<atom::internal, atom::persist>>::extend_with<
+    store_builder_actor>::unwrap;
+
+struct active_store_state {
+  filesystem_actor fs;
+
+  /// A pointer to the hosting actor.
+  //  We intentionally store a strong pointer here: The store lifetime is
+  //  ref-counted, it should exit after all currently active queries for this
+  //  store have finished, its partition has dropped out of the cache, and it
+  //  received all data from the incoming stream. This pointer serves to keep
+  //  the ref-count alive for the last part, and is reset after the data has
+  //  been written to disk.
+  local_store_actor self;
```

Comment:
It is true, but it shouldn't matter in both directions:

The upstream (ie. the index currently) will always close the stream stage before it is supposed to exit, so this should not extend the lifetime unnecessarily.

In the other direction, we don't want to rely on the ref-count from the upstream stream stage to keep the store alive, since that will usually drop to zero *before* all data from the stream arrived at the segment.